### PR TITLE
Do not include JSC and Hermes debug libraries when building in release mode

### PR DESCRIPTION
**Pull request by maxammann**
## Summary

Right now the `libhermes-executor-debug.so` and `libjscexecutor.so` get included when doing a release build with Hermes enabled. This can be fixed by deleting the .so files from the correct directories in the gradle build directory. They have to be deleted because they are packaged in the `./node_modules/react-native/android/com/facebook/react/react-native/0.63.2/react-native-0.63.2.aar` aar. This actually should be fixed somewhere else properly.

This PR fixes the workaround already implemented:

* Execute vmSelectionAction before strip${targetName}DebugSymbols instead of package${targetName}
* Remove .so files from merged_native_libs such that they don't get copied to stripped_native_libs

Before this PR multiple stack traces are thrown when starting an app:
```
11-02 19:46:36.413 24727 24768 W System.err: java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "_ZN8facebook6hermes13HermesRuntime11getDebuggerEv" referenced by "/data/app/~~ZHwF7KVKSeGGWCuyb8_i2Q==/tuerantuer.app.integreat-fSWEEZw7ODD62ZPjepvHlw==/lib/x86_64/libhermes-executor-debug.so"...
11-02 19:46:36.413 24727 24768 W System.err: 	at java.lang.Runtime.load0(Runtime.java:939)
11-02 19:46:36.413 24727 24768 W System.err: 	at java.lang.System.load(System.java:1628)
11-02 19:46:36.413 24727 24768 W System.err: 	at com.facebook.soloader.SoLoader$1.load(SoLoader.java:395)
11-02 19:46:36.413 24727 24768 W System.err: 	at com.facebook.soloader.DirectorySoSource.loadLibraryFrom(DirectorySoSource.java:77)
11-02 19:46:36.413 24727 24768 W System.err: 	at com.facebook.soloader.DirectorySoSource.loadLibrary(DirectorySoSource.java:50)
11-02 19:46:36.413 24727 24768 W System.err: 	at com.facebook.soloader.ApplicationSoSource.loadLibrary(ApplicationSoSource.java:82)
11-02 19:46:36.413 24727 24768 W System.err: 	at com.facebook.soloader.SoLoader.doLoadLibraryBySoName(SoLoader.java:766)
11-02 19:46:36.413 24727 24768 W System.err: 	at com.facebook.soloader.SoLoader.loadLibraryBySoName(SoLoader.java:673)
11-02 19:46:36.414 24727 24768 W System.err: 	at com.facebook.soloader.SoLoader.loadLibrary(SoLoader.java:611)
11-02 19:46:36.414 24727 24768 W System.err: 	at com.facebook.soloader.SoLoader.loadLibrary(SoLoader.java:559)
11-02 19:46:36.414 24727 24768 W System.err: 	at com.facebook.hermes.reactexecutor.HermesExecutor.<clinit>(HermesExecutor.java:22)
11-02 19:46:36.414 24727 24768 W System.err: 	at com.facebook.hermes.reactexecutor.HermesExecutorFactory.create(HermesExecutorFactory.java:29)
11-02 19:46:36.414 24727 24768 W System.err: 	at com.facebook.react.ReactInstanceManager$5.run(ReactInstanceManager.java:1017)
11-02 19:46:36.414 24727 24768 W System.err: 	at java.lang.Thread.run(Thread.java:923)
11-02 19:46:36.414 24727 24768 E SoLoader: couldn't find DSO to load: libhermes-executor-debug.so caused by: dlopen failed: cannot locate symbol "_ZN8facebook6hermes13HermesRuntime11getDebuggerEv" referenced by "/data/app/~~ZHwF7KVKSeGGWCuyb8_i2Q==/tuerantuer.app.integreat-fSWEEZw7ODD62ZPjepvHlw==/lib/x86_64/libhermes-executor-debug.so"... result: 0
11-02 19:46:36.414 24727 24768 D SoLoader: About to load: libhermes-executor-release.so
```
```
11-02 19:46:36.356 24727 24727 D SoLoader: About to load: libjscexecutor.so
11-02 19:46:36.356 24727 24727 D SoLoader: libjscexecutor.so not found on /data/user/0/tuerantuer.app.integreat/lib-main
11-02 19:46:36.356 24727 24727 D SoLoader: libjscexecutor.so found on /data/app/~~ZHwF7KVKSeGGWCuyb8_i2Q==/tuerantuer.app.integreat-fSWEEZw7ODD62ZPjepvHlw==/lib/x86_64
11-02 19:46:36.356 24727 24727 D SoLoader: Not resolving dependencies for libjscexecutor.so
11-02 19:46:36.356   307   307 W EmuHWC2 : No layers, exit, buffer 0x7a9456c0e970
11-02 19:46:36.358 24727 24727 W System.err: java.lang.UnsatisfiedLinkError: dlopen failed: library "libjsc.so" not found: needed by /data/app/~~ZHwF7KVKSeGGWCuyb8_i2Q==/tuerantuer.app.integreat-fSWEEZw7ODD62ZPjepvHlw==/lib/x86_64/libjscexecutor.so in namespace classloader-namespace
11-02 19:46:36.359 24727 24727 W System.err: 	at java.lang.Runtime.load0(Runtime.java:939)
11-02 19:46:36.359 24727 24727 W System.err: 	at java.lang.System.load(System.java:1628)
11-02 19:46:36.359 24727 24727 W System.err: 	at com.facebook.soloader.SoLoader$1.load(SoLoader.java:395)
11-02 19:46:36.359 24727 24727 W System.err: 	at com.facebook.soloader.DirectorySoSource.loadLibraryFrom(DirectorySoSource.java:77)
11-02 19:46:36.359 24727 24727 W System.err: 	at com.facebook.soloader.DirectorySoSource.loadLibrary(DirectorySoSource.java:50)
11-02 19:46:36.359 24727 24727 W System.err: 	at com.facebook.soloader.ApplicationSoSource.loadLibrary(ApplicationSoSource.java:82)
11-02 19:46:36.359 24727 24727 W System.err: 	at com.facebook.soloader.SoLoader.doLoadLibraryBySoName(SoLoader.java:766)
11-02 19:46:36.359 24727 24727 W System.err: 	at com.facebook.soloader.SoLoader.loadLibraryBySoName(SoLoader.java:673)
11-02 19:46:36.359 24727 24727 W System.err: 	at com.facebook.soloader.SoLoader.loadLibrary(SoLoader.java:611)
11-02 19:46:36.359 24727 24727 W System.err: 	at com.facebook.soloader.SoLoader.loadLibrary(SoLoader.java:559)
11-02 19:46:36.359 24727 24727 W System.err: 	at com.facebook.react.ReactInstanceManagerBuilder.getDefaultJSExecutorFactory(ReactInstanceManagerBuilder.java:297)
11-02 19:46:36.359 24727 24727 W System.err: 	at com.facebook.react.ReactInstanceManagerBuilder.build(ReactInstanceManagerBuilder.java:270)
11-02 19:46:36.359 24727 24727 W System.err: 	at com.facebook.react.ReactNativeHost.createReactInstanceManager(ReactNativeHost.java:87)
11-02 19:46:36.359 24727 24727 W System.err: 	at com.facebook.react.ReactNativeHost.getReactInstanceManager(ReactNativeHost.java:39)
11-02 19:46:36.359 24727 24727 W System.err: 	at tuerantuer.app.integreat.MainApplication.onCreate(MainApplication.java:45)
11-02 19:46:36.359 24727 24727 W System.err: 	at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1192)
11-02 19:46:36.359 24727 24727 W System.err: 	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6712)
11-02 19:46:36.359 24727 24727 W System.err: 	at android.app.ActivityThread.access$1300(ActivityThread.java:237)
11-02 19:46:36.359 24727 24727 W System.err: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1913)
11-02 19:46:36.359 24727 24727 W System.err: 	at android.os.Handler.dispatchMessage(Handler.java:106)
11-02 19:46:36.359 24727 24727 W System.err: 	at android.os.Looper.loop(Looper.java:223)
11-02 19:46:36.359 24727 24727 W System.err: 	at android.app.ActivityThread.main(ActivityThread.java:7656)
11-02 19:46:36.360 24727 24727 W System.err: 	at java.lang.reflect.Method.invoke(Native Method)
11-02 19:46:36.360 24727 24727 W System.err: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
11-02 19:46:36.360 24727 24727 W System.err: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
11-02 19:46:36.360 24727 24727 E SoLoader: couldn't find DSO to load: libjscexecutor.so caused by: dlopen failed: library "libjsc.so" not found: needed by /data/app/~~ZHwF7KVKSeGGWCuyb8_i2Q==/tuerantuer.app.integreat-fSWEEZw7ODD62ZPjepvHlw==/lib/x86_64/libjscexecutor.so in namespace classloader-namespace result: 0
11-02 19:46:36.360 24727 24727 D SoLoader: init exiting
```
Afterwards they are gone:
```
11-02 19:27:08.052 23251 23293 D SoLoader: About to load: libhermes-executor-debug.so
11-02 19:27:08.052 23251 23293 D SoLoader: libhermes-executor-debug.so not found on /data/user/0/tuerantuer.app.integreat/lib-main
11-02 19:27:08.052 23251 23293 D SoLoader: libhermes-executor-debug.so not found on /data/app/~~0mYQ5Jyf4NG9lK86d9nkjw==/tuerantuer.app.integreat-tyc3LBv82Ac1cO7aGgEP3w==/lib/x86_64
11-02 19:27:08.052 23251 23293 D SoLoader: libhermes-executor-debug.so not found on /vendor/lib
11-02 19:27:08.053 23251 23293 D SoLoader: libhermes-executor-debug.so not found on /system/lib
```

```
11-02 19:27:07.977 23251 23251 D SoLoader: About to load: libjscexecutor.so
11-02 19:27:07.977 23251 23251 D SoLoader: libjscexecutor.so not found on /data/user/0/tuerantuer.app.integreat/lib-main
11-02 19:27:07.978 23251 23251 D SoLoader: libjscexecutor.so not found on /data/app/~~0mYQ5Jyf4NG9lK86d9nkjw==/tuerantuer.app.integreat-tyc3LBv82Ac1cO7aGgEP3w==/lib/x86_64
11-02 19:27:07.979 23251 23251 D SoLoader: libjscexecutor.so not found on /vendor/lib
11-02 19:27:07.979 23251 23251 D SoLoader: libjscexecutor.so not found on /system/lib
```

## Changelog

* Android Fixed - Do not include JSC and Hermes debug libraries when building in release mode

## Test Plan

* Check for the output of `adb logcat` for the stacktraces. If they are gone after implementing this fix then it works. Run `./gradlew clean` every time before building.

### CODE REVIEW

File: react.gradle [hunk 1]
```diff
@@ -317,10 +317,12 @@ afterEvaluate {
...
         // This should really be done by packaging all Hermes releated libs into
         // two separate HermesDebug and HermesRelease AARs, but until then we'll
-        // kludge it by deleting the .so files out of the /transforms/ directory.
+        // kludge it by deleting the .so files out of the /merged_native_libs/ directory.
         def isRelease = targetName.toLowerCase().contains("release")
-        def libDir = "$buildDir/intermediates/transforms/"
+        def libDir = "$buildDir/intermediates/merged_native_libs"
         def vmSelectionAction = {
+            logger.error("$buildDir")
+            logger.error("vmSelectionActionvmSelectionAction")
```
Comment (maxammann) [inline react.gradle:325]:
```suggestion
```

File: react.gradle [hunk 2]
```diff
@@ -340,16 +340,14 @@ afterEvaluate {
                     include "**/libhermes*.so"
                 }
             }.visit { details ->
-                def targetVariant = ".*/transforms/[^/]*/${targetPath}/.*"
-                def path = details.file.getAbsolutePath().replace(File.separatorChar, '/' as char)
-                if (path.matches(targetVariant) && details.file.isFile()) {
+                if (details.file.isFile()) {
```
Comment (maxammann) [inline react.gradle:343]: Checking whether this is a file should be enough, as we are only visiting a single directory

**Comment by maxammann** [comment]
Fixes #29064 and https://github.com/facebook/hermes/issues/96

**Comment by maxammann** [comment]
A workaround which works right now is:
```

 if (enableHermes) {
            def taskName = getGradle().getStartParameter().getTaskRequests().toString()

            exclude "**/libjsc*.so"
            if (taskName.toLowerCase().contains("release")) {
                exclude '**/libhermes-inspector.so'
                exclude '**/libhermes-executor-debug.so'
            } else {
                exclude '**/libhermes-executor-release.so'
            }
        }
}
```

thanks to @prempalsingh

**Comment by ecreeth** [comment]
cc @fkgozali

**Comment by fkgozali** [comment]
There's a failure in FB internal CI for some reason, I'll need some time to look into that. CircleCI test_android seems green with this PR though.

**Comment by fkgozali** [comment]
@maxammann: it looks like something in this change doesn't play well with multiple flavors of rn-tester build. I think gradle got confused with the cached NDK build artifacts from the previous runs. To repro, we can try building with and without Fabric, and the 2nd build always failed, until you try again. For example:

```
# OK
./gradlew packages:rn-tester:android:app:packageHermesDebug

# Failed:
USE_FABRIC=1 ./gradlew packages:rn-tester:android:app:packageHermesDebug

# OK (2nd run, the same command as before)
USE_FABRIC=1 ./gradlew packages:rn-tester:android:app:packageHermesDebug

# Failed: if we disable USE_FABRIC again
./gradlew packages:rn-tester:android:app:packageHermesDebug

# OK (2nd run: the same command as before)
./gradlew packages:rn-tester:android:app:packageHermesDebug
```

This means, the removal logic in this PR incorrectly notifies gradle of the changes, causing incorrect caching of task outputs. Can you look into how to make each run consistently succeed without having to run the same command twice?

**Comment by maxammann** [comment]
@fkgozali thanks for checking the changes! I replaced the hacky way with a more stable way. I think this should solve problems with compiling multiple times

**Comment by maxammann** [comment]
Updating branch to rerun failed ios tests

**Comment by fkgozali** [comment]
The iOS failures seem unrelated, but test_android had:

```
1: Task failed with an exception.
-----------
* Where:
Build file '/root/react-native/packages/rn-tester/android/app/build.gradle' line: 86

* What went wrong:
A problem occurred evaluating script.
> No such property: name for class: java.lang.String
```

**Comment by maxammann** [comment]
@fkgozali The api I need is only available with the gradle plugin 4.2 which is still alpha: https://developer.android.com/reference/tools/gradle-api/4.2/com/android/build/api/variant/ApplicationVariant#packagingOptions(kotlin.Function1)

I think it makes sense to wait with this PR until we are at version 4.2. What do you think?

**Comment by maxammann** [comment]
Alternatively there is this solution: https://github.com/Jween/android-soexcluder/blob/master/app/src/main/groovy/com/jween/gradle/soexcluder/AppPlugin.groovy#L66

**Comment by fkgozali** [comment]
> I think it makes sense to wait with this PR until we are at version 4.2. What do you think?

I think this is probably best next step. We tend to upgrade AGP regularly, and if this is a good reason to upgrade, then it will make it easier for us to move forward during upgrade.

> Alternatively there is this solution

I'm hesitant to add a custom logic that's supposed to be supported by AGP directly, so unless there's a super urgent reason for excluding the .so right now, let's wait.

What do you think?

**Comment by maxammann** [comment]
Jup lets wait, I can close this and reopen it in the future. What do you think?

**Comment by fkgozali** [comment]
> Jup lets wait, I can close this and reopen it in the future. What do you think?

Sure, thanks!

**Comment by maxammann** [comment]
Gradle 4.2 is released now and this could be fixed. I'll also check whether this still happens. Does RN already use 4.2?

**Comment by maxammann** [comment]
Yet it is :) https://github.com/facebook/react-native/blob/master/build.gradle.kts#L15

**Comment by cortinico** [comment]
I'm closing this as we now have a different mechanism to exclude unnecessary libs from being bundled in the final .apk/.aab

**Comment by rkok** [comment]
@cortinico Could you place a link to where that new mechanism is explained? Thanks in advance.

**Comment by cortinico** [comment]
@rkok yup here it is:
https://github.com/facebook/react-native/blob/ea29ae1ceb3845e8116e2e9cf9399ab38523b777/react.gradle#L409-L464

**Pull request closed without merging**